src/BitBucketOnPrem.ts

```diff
@@ -135,12 +195,132 @@ export const getFileTreeFromBitbucket =
         return files;
     };
 
+export const cacheFileTree = async ({url, accessToken, projectKey, repoName, commit}: BitbucketOnPrem): Promise<boolean> => {
+    // If a repo is currently being cached, don't cache another one (log a warning and return false)
+    if (repoCurrentlyBeingCached) {
+        logger.warn("Cannot cache two repos at the same time");
+        return false;
+    }
+    try {
+        // This allows to query explorook about the repo currently being cached
+        repoCurrentlyBeingCached = {projectKey, repoName, commit};
+        const fileTreeUrl = UrlAssembler(url).template("/rest/api/1.0/projects/:projectKey/repos/:repoName/files").param({
+            projectKey,
+            repoName
+        }).query({
+            at: commit
+        }).toString();
+
+        logger.debug("Getting files for", {projectKey, repoName, url, commit});
+        let isLastPage = false;
+        let start = 0;
+        let files: string[] = [];
+        const limit: number = await getFileTreePageLimit({url, accessToken, projectKey, repoName, commit});
+        while (!isLastPage) {
+            const filesBatch = await fetchTreeParallel({fileTreeUrl, accessToken, start, limit});
+            if (abortCache) {
+                logger.debug("Cache aborted via api");
+                return false;
+            }
+            files = files.concat(filesBatch[0], filesBatch[1], filesBatch[2], filesBatch[3], filesBatch[4]);
+            if (filesBatch[0]?.length && filesBatch[1]?.length && filesBatch[2]?.length && filesBatch[3]?.length && filesBatch[4]?.length) {
```

Comment:
maybe move this to a function that iterates? and check that all items have length

Reply:
I think having the function called "validateAllBatchesAreFull()"  or something like that it will be readable enough
I don't see this affected performance
and sure keep it with the hard coded numbers

Reply:
this is really an IMO
we can keep it as it is if you want

Reply:
> this is really an IMO we can keep it as it is if you want

I am testing for performance diff now

Reply:
> I think having the function called "validateAllBatchesAreFull()" or something like that it will be readable enough I don't see this affected performance and sure keep it with the hard coded numbers

@hopemanryan Apparently it is even faster when I flatten it, so I avoided the repeat :)